File: verifier-cli/src/main/java/com/jetbrains/pluginverifier/verifiers/clazz/InheritFromFinalClassVerifier.java
package com.jetbrains.pluginverifier.verifiers.clazz;

import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.location.ProblemLocation;
import com.jetbrains.pluginverifier.problems.ClassNotFoundProblem;
import com.jetbrains.pluginverifier.problems.InheritFromFinalClassProblem;
import com.jetbrains.pluginverifier.verifiers.VerificationContext;
import com.jetbrains.pluginverifier.verifiers.util.VerifierUtil;
import org.objectweb.asm.tree.ClassNode;

public class InheritFromFinalClassVerifier implements ClassVerifier {
  public void verify(final ClassNode clazz, final Resolver resolver, final VerificationContext ctx) {
    final String superClassName = clazz.superName == null ? "java/lang/Object" : clazz.superName;
    ClassNode supClass = VerifierUtil.findClass(resolver, superClassName, ctx);
    if (supClass == null) {
      if (!ctx.getVerifierOptions().isExternalClass(superClassName)) {
        ctx.registerProblem(new ClassNotFoundProblem(superClassName), ProblemLocation.fromClass(clazz.name));
      }
      return;
    }
    if (VerifierUtil.isFinal(supClass)) {
      ctx.registerProblem(new InheritFromFinalClassProblem(supClass.name), ProblemLocation.fromClass(clazz.name));
    }
  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/clazz/InheritFromFinalClassVerifier.kt
package com.jetbrains.pluginverifier.verifiers.clazz

import com.jetbrains.pluginverifier.results.problems.InheritFromFinalClassProblem
import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.createClassLocation
import com.jetbrains.pluginverifier.verifiers.isFinal
import com.jetbrains.pluginverifier.verifiers.resolveClassOrProblem
import org.objectweb.asm.tree.ClassNode

class InheritFromFinalClassVerifier : ClassVerifier {
  override fun verify(clazz: ClassNode, ctx: VerificationContext) {
    val superClassName = clazz.superName ?: return
    val supClass = ctx.resolveClassOrProblem(superClassName, clazz, { clazz.createClassLocation() }) ?: return
    if (supClass.isFinal()) {
      val child = clazz.createClassLocation()
      val finalClass = supClass.createClassLocation()
      ctx.registerProblem(InheritFromFinalClassProblem(child, finalClass))
    }
  }
}
